### DISCUSSION TITLE
Storefront API: How to use Shopify-Storefront-Private-Token?

### DISCUSSION THREAD

**Discussion by jesperp** [Shopify/storefront-api-feedback#126] [Help] [answered]
In the docs here it says we can use a "delegate access token" in conjunction with headers `Shopify-Storefront-Private-Token` and `Shopify-Storefront-Buyer-IP` to deal with rate limiting while in a server environment.

But this doesn't work. Several people have tried this and failed:

https://community.shopify.com/c/shopify-apis-and-sdks/storefront-api-generation-of-delegate-access-token-shopify/m-p/1650815?search-action-id=95471651770&search-result-uid=1650815
https://community.shopify.com/c/storefront-api-and-sdks/403-response-when-using-shopify-storefront-private-token/m-p/1641032?search-action-id=95471651770&search-result-uid=1641032
https://community.shopify.com/c/storefront-api-and-sdks/how-to-avoid-rate-limit-when-using-server-side-rendering-ssr/m-p/1671021?search-action-id=95471651770&search-result-uid=1671021

How do we create a token to pass into the `Shopify-Storefront-Private-Token` header? I have tried both [delegateAccessTokenCreate](https://shopify.dev/api/admin-graphql/2022-07/mutations/delegateAccessTokenCreate) and [storefrontAccessTokenCreate](https://shopify.dev/api/admin-graphql/2022-07/mutations/storefrontAccessTokenCreate)  but I still get 403s when trying to use the tokens. At first I thought `Shopify-Storefront-Private-Token` was a typo (where is the `X-`?), but even with the prefix it is still not working.

⚠️  This is a serious blocker for us and would really appreciate a swift response!

**Comment by johncraigcole** [comment, marked as answer]
Hi @jesperp 👋. Thanks for your question about this. We're rolling this out incrementally. The first priority was for Hydrogen, and is also why you're able to find general details in our docs. Once we're at 100% rollout we'll update the [developer changelog](https://shopify.dev/changelog), and provide additional documentation on how to proper utilize the request headers you've identified. I don't have an exact timeline to share today, and while things can change, I feel confident that you'll see this be fully available later this quarter.

**Reply by andresespinosapc** [reply]
And how are we supposed to use the Storefront API without reaching the rate limit?

**Reply by johncraigcole** [reply]
If you're using a server rendered solution that is making Storefront API requests server side outside of a Hydrogen context yes you will hit API rate limits. 

When the server to server flow is released for non Hydrogen contexts, you'll leverage the new request headers `Shopify-Storefront-Buyer-IP` and `Shopify-Storefront-Private-Token` to pass the buyer requesting IP for which the rate limit will be based and should ensure you don't run into any API rate limits. I'm still hopeful this will be available towards the end of this quarter barring any unforeseen changes that shift priority.

**Reply by andresespinosapc** [reply]
Then you should remove the `Usage limitations` section from the [rate limit's documentation](https://shopify.dev/api/storefront#rate_limits), I built a complete app thinking I could use server-side rendering and now I stumble upon this.

**Reply by andresespinosapc** [reply]
Last question @johncraigcole , when this is fully available it will be for all accounts or only Plus ones?

**Reply by Gldywn** [reply]
I totally agree @andresespinosapc, the documentation is totally misleading..
I'm in the same situation, @johncraigcole is it possible to share with us some quick guidelines on how to use this header to unblock us?

**Reply by jesperp** [reply]
@andresespinosapc @0xWhiteleaf Please read the responses from @johncraigcole again. The answers are there

**Reply by Gldywn** [reply]
@jesperp @johncraigcole I've already read this but it doesn't unlock the situation for people who followed the documentation and thought they could use Storefront API from a back-end. I just ask for two or three quick lines to use it the same way Oxygen does?

**Reply by jesperp** [reply]
As John said this is only for Hydrogen users initially. Rollout for all users expected later this quarter

**Reply by johncraigcole** [reply]
As @jesperp has indicated this was only released initially for Hydrogen. At that time doc updates were made in support of that initiative which ultimately has caused some confusion as to what is currently possible in a server to server flow outside of a Hydrogen context. I apologize for the confusion and impact the doc updates had in this instance. That being said, we're working towards getting this functionality available to all server to server flows as soon as we can.

**Reply by johncraigcole** [reply]
@andresespinosapc I confirmed that once fully released the server to server flow will work for any plan that can use the Storefront API. So basic right up to Plus will be supported.

**Reply by drew-foxall** [reply]
@johncraigcole I think it would be worth clarifying what the end of the quarter means to Shopify. 

The documentation changes were made in July and we are in mid-September now. 

As Shopify partners working with Headless clients, this puts us in a challenging position. It took well over a month to receive any sort of official word on what the documentation meant and that it was, for the time being at least, incorrect.

Now that developers and partners have active development projects based on incorrect documentation, we urgently need clarity.

**Reply by johncraigcole** [reply]
Hey @Drew-Garratt we don't typically give out launch dates for features. Given there was some confusion over what was possible outside of a Hydrogen context I wanted to give the community a rough idea of when the server to server flow would be available outside of a Hydrogen context. My statement about the end of the quarter was to provide general guidance on when the feature would be fully released. Put another way, we hope to have this feature fully available for all server to server flows by the end of September 2022. The moment it's released I'll be here to share that with the community.

**Reply by drew-foxall** [reply]
@johncraigcole Thank you for the response and the clarification. Typically on features in any large-scale product like this, I wouldn't push. The need here has only really arisen because the documentation still leads developers and merchants alike to pursue solutions that aren't yet available. Because of that, I think many of us found ourselves needing an official word on when the situation could be resolved. So, once again, you have my thanks.

**Reply by johncraigcole** [reply]
Hey Everyone we're now live 🎊 🎉 . [Change log entry](https://shopify.dev/changelog/use-a-server-to-build-custom-storefronts-with-authenticated-access-for-the-storefront-api) is published with link to improved help docs. Thanks for all the patience as we brought the server to server flow to all use cases.

**Reply by pbright** [reply]
@johncraigcole Thanks for this change. Are the rate limits when using authenticated access with the Storefront API published anywhere? I couldn't find it in the [Storefront API docs](https://shopify.dev/api/storefront#rate_limits) or in the [generic API usage rate limits](https://shopify.dev/api/usage/rate-limits) section.

**Reply by johncraigcole** [reply]
The rate limiting will be the same as documented [here](https://shopify.dev/api/usage/rate-limits#storefront-api-rate-limits). Storefront API is a time based system by app and ip. If you have any questions about the information in the link I provided let me know.

**Reply by pbright** [reply]
Thanks @johncraigcole. 

The change log states:

> Using authenticated access enables more throughput for your server than using a public token...

Additionally, the [docs for authenticated access](https://shopify.dev/api/usage/authentication#getting-started-with-authenticated-access) (linked to from the change log entry) state:

> With authenticated access, your requests are throttled at the shop level and optionally by a forwarded IP. This is advantageous because your server needs a much larger capacity for making requests for many buyers. The app throttle scales with the Shopify platform and isn't fixed. Under high load, such as for flash sales, both the platform and the throttle scale to support more requests.

I took those two statements to mean that the rate limit would be increased when using authenticated access, which would help in server environments where all requests are made to Shopify from the server IP address. We will use the `Shopify-Storefront-Buyer-IP`  header where practical, although there are situations where that is impossible (such as during static site generation) and the docs make it clear that header is optional.

The rate limit docs you linked to state:

> The 60-second limit applies to the IP address of the buyer interacting with the app, rather than the ID of the store where the app is installed

That seems contradictory with the authenticated access snippet above which states they are throttled at the shop level, and there is no separate mention in those docs of rate limits when using authenticated access.

Is there any change to the rate limits when using authenticated access?

**Reply by johncraigcole** [reply]
Hi @pbright thanks for the detailed follow up. They'll be more info and content coming on this topic but let me try to provide a high level overview in the meantime.

Please continue to use the `Shopify-Storefront-Buyer-IP` header when it's a buyer act that kicked off a request. This helps ensure that no one specific IP, including potential bot traffic, is permitted to use up too much capacity. And yes, when the `Shopify-Storefront-Buyer-IP` is provided, the limiting will be based on time and per ip.

If the `Shopify-Storefront-Buyer-IP` is not provided, the Storefront API limit based on time and ip will not apply. Not providing the buyer ip header has specific use cases. For example, a static site build where a burst of requests are required to generate the site. With an IP time based limit this would likely lead to API rate limiting. Without the `Shopify-Storefront-Buyer-IP` header there isn't any traditional Shopify API rate limits but there are still platform protections in place. If platform health ever reaches a critical point, requests can be dropped so that requests of higher priority (like access to checkout) can continue. When not passing the `Shopify-Storefront-Buyer-IP` your solution should expect the possibility of requests being dropped and should implement an exponential backoff. Again, the inclusion of the `Shopify-Storefront-Buyer-IP` header is what helps ensure heavy traffic from a given ip can be limited and offer protection to the merchant and buyers interacting with the custom storefront.

**Reply by pbright** [reply]
Thanks @johncraigcole, really appreciate the follow up. That all makes sense and sounds like exactly how we were hoping it would work for our use case.

**Comment by patjakubik** [comment]
Hey @johncraigcole I have additional question about delegate access tokens. Are those usable for the public apps that aren't sales channels? I added some unauthenticated scopes to my app and was able to successfully create a delegate access token. But the token itself seems "broken". I'm unable to query anything (getting INTERNAL_SERVER_ERROR), and introspection doesn't work either. That worked for another project, which is a custom sales app. The requestId of mentioned error is: f4b3a26e-18c6-455c-a3e1-1d8f39b9ffc8

**Reply by johncraigcole** [reply]
Hey @patryk-smc 👋. The app token you used to generate the delegate was it an offline or online token? Also could you confirm the request headers you're sending when making the Storefront API request.

I'm not seeing the request-id provided in our logs. If you could, would you mind running another test and provide me with a new request id. 

Thanks

**Reply by patjakubik** [reply]
@johncraigcole I used offline token. Here's another text:

Request timeline:

```
> POST /api/2022-10/graphql.json HTTP/2
> Host: ***redacted***.myshopify.com
> user-agent: insomnia/2022.6.0
> shopify-storefront-private-token: shpat_1baa6b***redacted***
> content-type: application/json
> accept: */*
> content-length: 51

| {"query":"query {\n\tshop {\n\t\tname\n\t}\n}\n\n"}

* We are completely uploaded and fine

< HTTP/2 200 
< date: Fri, 14 Oct 2022 20:05:19 GMT
< content-type: application/json; charset=utf-8
< x-sorting-hat-podid: 131
< x-sorting-hat-shopid: 37402968196
< x-storefront-renderer-rendered: 1
< vary: Accept-Encoding
< vary: Accept
< x-frame-options: DENY
< content-security-policy: block-all-mixed-content; frame-ancestors 'none'; upgrade-insecure-requests;
< access-control-allow-origin: *
< x-shopify-api-version: 2022-10
< x-shopid: 37402968196
< x-shardid: 131
< content-language: pl
< server-timing: processing;dur=40, db;dur=23
< x-shopify-stage: production
< x-dc: gcp-europe-west3,europe-west1,gcp-europe-west1
< x-request-id: cb56c2be-ef03-4b23-992a-cad5bae9b43c
< x-download-options: noopen
< x-xss-protection: 1; mode=block
< x-permitted-cross-domain-policies: none
< x-content-type-options: nosniff
< cf-cache-status: DYNAMIC
< report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=qmqH%2BAintqUhYU%2FOLQySLFZXxNrXwvzULxV6sIj3fbWFSsKcpSzqXvJ521XkNZVz1O6jQroubfp4%2F7E%2BwqQge%2FmKXAgFMBgn3zIA16wrVx0AtyZNTa%2BCeHmHOiJvi55dIacTn1sFnmxGkYHi75%2Bu6PQ%3D"}],"group":"cf-nel","max_age":604800}
< nel: {"success_fraction":0.01,"report_to":"cf-nel","max_age":604800}
< server: cloudflare
< cf-ray: 75a2eddb1c50bfb7-WAW
< alt-svc: h3=":443"; ma=86400, h3-29=":443"; ma=86400
```

Response body:

```
{
	"errors": [
		{
			"message": "Internal error. Looks like something went wrong on our end.\nRequest ID: cb56c2be-ef03-4b23-992a-cad5bae9b43c (include this in support requests).",
			"extensions": {
				"code": "INTERNAL_SERVER_ERROR",
				"requestId": "cb56c2be-ef03-4b23-992a-cad5bae9b43c"
			}
		}
	]
}
```

**Reply by johncraigcole** [reply]
Hey @patryk-smc I think this could be just in the scope being asked for. From the query I can see that you're reading the shop name using the Storefront API. In trying to reproduce the problem, I found that if I created a delegate token with only `unauthenticated_read_checkouts` I got an error. I'm not sure if that is the only scope you setup with the delegate but would be curious if it was. Assuming for a second it was and if your public app instead asked for `unauthenticated_read_product_listings` and generated a delegate for that scope would the shop query work?

**Reply by patjakubik** [reply]
Hey @johncraigcole that's great idea but unfortunately did not help. I get the same error. See request ID: 40222d36-1df9-4da4-8a4d-511e1d3e94d3

I also tried querying products. The error is different now. Instead, I get HTTP 500 Internal Server Error. See request ID: 62f10e7c-a8fb-40d9-b914-d7bb6817a63b

**Reply by patjakubik** [reply]
And answering your question - yes, my app was requesting `unauthenticated_read_checkouts` only and that's the scope I passed to delegateAccessTokenCreate mutation as well.

**Comment by johncraigcole** [comment]
Hey @patryk-smc 👋 thanks for the most recent request ids. The second of the two request-id's is pointing to a channel error. Is your  public app an actual sales channel?

**Reply by patjakubik** [reply]
Hey John, no my app is not a sales channel

**Reply by johncraigcole** [reply]
Sorry just realized you pretty much indicated that with the original ask 🤦. Do the SFAPI queries using the public app token work (not the delegate)?

**Reply by patjakubik** [reply]
No worries! Yep they do

**Reply by patjakubik** [reply]
No sorry I was wrong. I was unable to create public token

**Comment by lxnre-codes** [comment]
Hi @johncraigcole  according to this [dev changelog](https://shopify.dev/changelog/storefront-api-is-now-available-to-non-channel-apps), it says none sales channels public app can now request `storefront access token`. While trying to get the access token i keep getting this error "App must be a channel or have unauthenticated scopes to create a storefront access token", even after requesting unauthenticated scopes. Here's the request id `ce92b5d7-c350-4ea7-9b39-47435893cdff` if this helps

**Reply by johncraigcole** [reply]
Hey @0x-buidl 👋. I suspect your app is requesting `unauthenticated_write_checkouts` on install. This only historically works for sales channels which I don't believe your use case is. Instead, remove  `unauthenticated_write_checkouts` from the scopes your app asks for originally, and instead use the other scopes necessary. For example,  `unauthenticated_read_product_listings`.

**Reply by lxnre-codes** [reply]
Thanks @johncraigcole, just got it working by changing the scopes as suggested. Another issue here, i realized the `cart id`  is required either fetch / mutate the cart from storefront api. Is there any way to get the cart id from the regular `online store sales channel`. Most especially to add / remove discount codes and attributes. Figured i wouldn't be able to use the storefront api this way without the `cart id`.

**Reply by johncraigcole** [reply]
Technically today there's actually two carts. One for liquid, and another for the Storefront API. This is something we're working to address but can't share a timeline on when a single unified cart will be available. Generally speaking, the cart id can be found in a client side accessible cookie called `cart`. I'm just not sure when we have a unified cart if an app would be able to grab the cart id and start mutating or querying that cart with the Storefront API. 

What you have today is the ability to create a cart with the Storefront API and send that buyer to checkout but that checkout would be disconnected from the online store cart.

**Reply by lxnre-codes** [reply]
Thanks for clarifying.

**Reply by keiraarts** [reply]
Sorry for jumping in! 

I have the exact same issue as 0x-buidl but I can't seem to resolve it the same way.

1 - I removed all my unauthenticated scopes except for "unauthenticated_read_product_listings" and updated my access token.

2 - If I use the token with "Shopify-Storefront-Private-Token" I get the error "Channel not found for the given token.".

3 - If I use the token wth a "storefrontAccessTokenCreate" mutation  I get 'Access denied for storefrontAccessTokenCreate.".

The request-id is 5b07bf43-6cb1-4b8c-8a08-5f4c2f0152ba and 38445fa4-f339-4595-99e3-44ed7c93620b


`{
					"handle": "write_orders",
					"description": "Modify orders, transactions, and fulfillments"
				},
				{
					"handle": "write_order_edits",
					"description": "Modify orders"
				},
				{
					"handle": "read_customers",
					"description": "Read customer details and customer groups"
				},
				{
					"handle": "read_products",
					"description": "Read products, variants, and collections"
				},
				{
					"handle": "write_merchant_managed_fulfillment_orders",
					"description": "Action fulfillment for merchant-managed locations"
				},
				{
					"handle": "write_assigned_fulfillment_orders",
					"description": "Modify assigned fulfillment"
				},
				{
					"handle": "write_draft_orders",
					"description": "Modify draft orders"
				},
				{
					"handle": "write_shipping",
					"description": "Modify shipping rates, countries, and provinces"
				},
				{
					"handle": "unauthenticated_read_product_listings",
					"description": "Read product information"
				},
				{
					"handle": "read_orders",
					"description": "Read orders, transactions, and fulfillments"
				},
				{
					"handle": "read_order_edits",
					"description": "Read order edits"
				},
				{
					"handle": "read_merchant_managed_fulfillment_orders",
					"description": "Read fulfillment for merchant-managed locations"
				},
				{
					"handle": "read_assigned_fulfillment_orders",
					"description": "Read assigned fulfillment"
				},
				{
					"handle": "read_draft_orders",
					"description": "Read draft orders"
				},
				{
					"handle": "read_shipping",
					"description": "Read shipping rates, countries, and provinces"
				}`
				
Been at it for days and don't understand what could be missing!
